Update OSPRs with GitHub activity

gsong opened the pull request:
## Highlights

* Listen for the following GitHub PR events, and post updates to JIRA issues accordingly:
    * issue_comment
    * pull_request
    * pull_request_review
    * pull_request_review_comment
* Establish a new pattern for the bot:
    * Flask views do not need to know how to access GitHub and JIRA.
    * OAuth credentials are stored in environment variables, not in the database.
    * Flask views and asynchronous tasks do not share bootstrap process and environments.
    * Use [RQ](http://python-rq.org) instead of Celery for job queueing and processing.
    
    The general workflow is `views` → `dispatcher` → `tasks`.
    
## Deployment

When you’re ready for deployment, follow the steps in the HOWTO-activate-jira-github-activity.rst.

/cc @edx/ospr @wesmason @kylecrawshaw

nedbat commented on openedx_webhooks/services/jira/utils.py:
```diff
@@ -0,0 +1,49 @@
+    return jira.search_issues(jql)
+
+
+def make_edx_action_choices():
```
This function (and the constant in services/jira/constants.py) seems really tightly coupled to the update_latest_github_activity handler. They'd be easier to understand if you put this function into jira/tasks/update_latest_github_activity.py

gsong replied:
Agreed.

nedbat commented on requirements/base.in:
```diff
@@ -5,19 +5,24 @@ Flask-SQLAlchemy==2.0
 Flask-SSLify
 Flask-Script
 Flask==0.10.1
+PyGithub
```
I thought we had talked about using github3 as a python wrapper for GitHub?

nedbat commented on openedx_webhooks/jira/tasks/update_latest_github_activity.py:
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
```
I'm not sure the "jira/tasks" hierarchy is helping us here.  How is this task more about jira than it is about github? It has jira in the directory and github in the file name.  This code might be over-hierarchialized.

gsong replied:
This task *only* updates JIRA issues, it has no awareness of GitHub since it only receives static information.

nedbat replied:
The word "GitHub" is in the name!

gsong replied:
1. In this case `github` in the name refers to JIRA fields we're using to track GitHub activities.
2. Separation of concerns. This way we know that this set of tasks *only* interact with JIRA.

gsong commented on openedx_webhooks/decorators.py:
```diff
@@ -0,0 +1,54 @@
+"""
```
I didn't find anything readily. The only thing I know of is `django.utils.functional.lazy`, but maybe you have suggestions?

nedbat replied:
Searching PyPI under the "right" name ( :) ) finds a number of options: https://pypi.python.org/pypi?%3Aaction=search&term=cache%20decorator&submit=search

gsong replied:
Again, not about caching, but lazy evaluation.

gsong commented on openedx_webhooks/github/handlers/bouncer/__init__.py:
```diff
@@ -0,0 +1,31 @@
+from ...models import GithubHeader
+from .rules import RULES
+
+logger = logging.getLogger(__name__)
```
We can definitely remove it for now. My intention was to leave them there and go back to set up logging if we can agree on this design pattern. But we can definitely remove now and add logger in later.

nedbat commented on openedx_webhooks/services/edx_repo_tools_data/__init__.py:
```diff
@@ -0,0 +1,22 @@
+
+
+@lazy
+def _get_people():
```
This seems dangerous to make lazy, since it means the file will only be read once per process.  How will updates to the file be read? (BTW: this is why i dislike the common name "lazy" for this. It isn't about being lazy, all functions are lazy. It's about caching the result and only calling the function once.)

nedbat commented on openedx_webhooks/github/handlers/bouncer/__init__.py:
```diff
@@ -0,0 +1,31 @@
+        if rejected:
+            return message
+
+    q.enqueue('openedx_webhooks.github.handlers.process', event_type, event)
```
You get all three of those benefits from functions (or classes with methods).  Queueing is for other things (doing work on other machines, deferring work, responding quickly in spite of long work, retrying).  We don't need any of those things on the second and third queuing here.

gsong commented on openedx_webhooks/github/handlers/bouncer/__init__.py:
```diff
@@ -0,0 +1,31 @@
+# -*- coding: utf-8 -*-
+from __future__ import (
+    absolute_import, division, print_function, unicode_literals
+)
```
Changed.

gsong commented on openedx_webhooks/jira/tasks/update_latest_github_activity.py:
```diff
@@ -0,0 +1,33 @@
+
+import logging
+
+from ...services.jira import jira as _jira
```
Corrected.

gsong commented on openedx_webhooks/services/edx_repo_tools_data/models/person.py:
```diff
@@ -0,0 +1,64 @@
+    @property
+    def institution(self):
+        data = self.person.get('institution')
+        return self.has_agreement and data
```
Changed.

gsong commented on openedx_webhooks/services/edx_repo_tools_data/models/person.py:
```diff
@@ -0,0 +1,64 @@
+
+    @property
+    def is_edx_user(self):
+        is_edx = self.institution.lower() == 'edx'
```
Another good catch. Thank you.

gsong commented on openedx_webhooks/jira/tasks/utils.py:
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
```
Maybe, I don't know yet. We don't have any JIRA handlers yet. If it turns out there are useful utilities to be used by both tasks and handlers, then we can move them up to `jira.utils`.

gsong commented on openedx_webhooks/services/edx_repo_tools_data/models/person.py:
```diff
@@ -0,0 +1,64 @@
+logger = logging.getLogger(__name__)
+
+
+class Person(object):
```
Is this your preference or is this part of general Python/edX style preference?

nedbat commented on openedx_webhooks/jira/tasks/update_latest_github_activity.py:
```diff
@@ -0,0 +1,48 @@
+
+JIRA_FIELDS_ID_LOOKUP = make_fields_lookup(JIRA_PR_FIELDS._asdict().values())
+
+JIRA_EDX_ACTION_CHOICES = make_edx_action_choices()
```
I would put this function right here, in this file. It's not a general purpose utility, it's only useful to this task.

nedbat commented on openedx_webhooks/github_views.py:
```diff
@@ -5,35 +5,58 @@
 
 
+@github_bp.route('/hook-receiver', methods=('POST',))
+def hook_reciver():
```
`hook_receiver`

gsong commented:
@edx/ospr All of the concerns, with the exception of the GitHub API client, has been addressed:

* [x] Renamed `services` → `lib`
* [x] Changed from bouncer to dispatcher model
* [x] Instead of successively queueing the tasks, invoke the functions directly

I've also simplified the folder structures, thereby eliminating some of the abstractions (which we can put back when necessary).

I've also marked modules that we can unit test.

nedbat commented:
"I've also marked modules that we can unit test":  Can't we unit test all the modules? And shouldn't we?

gsong commented:
@nedbat I've marked ones that we should start with.

nedbat commented on .gitignore:
```diff
@@ -1,7 +1,9 @@
+!openedx_webhooks/lib/
```
Just curious: why did we need this line in the .gitignore?

nedbat replied:
Oh, because your .gitignore_global ignores "lib".  That seems really dangerous as a global setting.  I use the global gitignore only for unusual things that I personally add to repos.

nedbat replied:
I would remove this line from here, and the lib line from your global gitignore. Any repo that produces a lib directory as a build artifact should git-ignore it in the repo instead.

nedbat commented:
At this point, the code is fine to merge. We need to understand how noisy this will be before we can merge it.

gsong commented:
@nedbat I will create a release branch and merge there. That way we can keep track of additional changes more easily. Sound good?

nedbat commented:
@gsong My point was that we aren't sure we want to release this code yet.  The way edX typically manages its repos, changes are merged to master when they are ready to be released.

gsong commented:
@nedbat I understand that we're not releasing yet. Do you prefer additional changes to stay within this PR, or as separate PR so they're easier to review?

nedbat commented:
We've got a ton of changes already, so we might as well keep going here.  What kinds of changes?

nedbat commented on bin/update_or_create_webhooks.py:
```diff
@@ -0,0 +1,59 @@
+    Update or create the specified webhook into repos using a
+    configuration JSON file:
+
+    \b
```
What is `\b` here?

nedbat replied:
Bleh: http://click.pocoo.org/5/documentation/#preventing-rewrapping   

Click couldn't have used some existing .rst syntax? Ugh. Or if they're going to make up something of their own, use something readable instead of a backspace character.... Carry on.

nedbat replied:
I don't understand where this JSON file is coming from? We already had repos.json, and that was replaced with openedx.yml files in the repos themselves.  Can't we use the openedx.yml files instead of creating a new configuration file?

gsong replied:
These are runtime arguments for the CLI, we could inline the arguments, but they're too complex to easily express that way.

We could store the webhook specs in `edx/repo-tools-data`. Let's discuss what you have in mind so I can understand your viewpoint better.

nedbat commented on openedx_webhooks/lib/github/utils.py:
```diff
@@ -0,0 +1,92 @@
+def get_webhooks(repo, payload_url):
+    # TODO: Test
+    hooks = []
+    for hook in (hook for hook in repo.iter_hooks() if hook.name == 'web'):
```
I would put this on two lines.  Looping over a generator expression with a conditional in it seems too tricky, especially since there's a loop body here anyway, with an if in it.

nedbat commented on openedx_webhooks/lib/github/utils.py:
```diff
@@ -0,0 +1,92 @@
+        if repo_contains_webhook(repo, payload_url, include_not_active):
+            repos.append(repo)
+
+    return set(repos)
```
Why make this into a set?  The list seems like exactly what you want. There are no duplicates, right?

nedbat commented on openedx_webhooks/lib/github/utils.py:
```diff
@@ -0,0 +1,92 @@
+
+def _get_most_recent_hook(hooks):
+    sorted_hooks = sorted(hooks, lambda h: h.updated_at, reverse=True)
+    return sorted_hooks[0] if sorted_hooks else None
```
Looks like you want min() instead of sorted()

nedbat commented on bin/update_or_create_webhooks.py:
```diff
@@ -0,0 +1,80 @@
+            '**Dry run only** The following actions would have been performed:'
+        )
+
+    for repo, conf in ((r, c) for r in repos for c in hook_confs):
```
This would be better as a double loop instead of a for statement over a double generator expression.

gsong replied:
Simply for legibility?

nedbat commented on HOWTO-activate-jira-github-activity.rst:
```diff
@@ -0,0 +1,47 @@
+   -  [ ] JIRA\_OAUTH\_PRIVATE\_KEY
+   -  [ ] JIRA\_SERVER
+
+2. Add ``rqwoker`` dyno type
```
"rqworker"

nedbat commented on HOWTO-activate-jira-github-activity.rst:
```diff
@@ -0,0 +1,47 @@
+   ``bin/gh_list_repos_with_webhook.py`` command.
+
+4. Add the following webhook configuration to
+   ``openedx_webhooks.webhook_confs``:
```
This has to be edited into a .py file, right? I looked in openedx_webooks/webhook_confs.py, and it wasn't clear to me how to combine what's there with this code.

nedbat commented on HOWTO-activate-jira-github-activity.rst:
```diff
@@ -0,0 +1,47 @@
+               'pull_request_review',
+               'pull_request_review_comment',
+           ]
+       }
```
Then there's a final step, right? Re-deploy on heroku?

nedbat commented on HOWTO-activate-jira-github-activity.rst:
```diff
@@ -0,0 +1,47 @@
+       heroku ps:scale rqworker=1
+
+3. Install the webhooks at various repos by using the
+   ``bin/gh_list_repos_with_webhook.py`` command.
```
Where do I run this command? Does it work if I do it locally?

nedbat commented on openedx_webhooks/github/dispatcher/rules/__init__.py:
```diff
@@ -0,0 +1,29 @@
+    github_activity,
+)
+"""
+Tuple[Module, ...]: A tuple of rules to process, in order
```
Is this the new typing syntax? I see these docstring-like strings, which aren't docstrings. Why aren't they comments?

Personally, I'd prefer to use a list for RULES and things like it.  Lists are like arrays, tuples are like structs.

nedbat commented on openedx_webhooks/webhook_confs.py:
```diff
@@ -0,0 +1,23 @@
+    'events': ['pull_request']
+})
+"""
+Tuple[Dict[
```
The code above isn't a tuple of dicts. It's a parenthesized dict. (Lists are better for these anyway :)  )

nedbat commented on openedx_webhooks/webhook_confs.py:
```diff
@@ -0,0 +1,23 @@
+})
+"""
+Tuple[Dict[
+    str, Union[Dict[str, str], List[str]]
```
`Dict[str, str]` here doesn't capture the type of what you have above. Two of your keys map to strings, but one is a boolean.  I'm not sure this type syntax is helping us much. Better would be to describe the keys in the dict, what they mean, and what value they expect.

gsong commented:
@edx/ospr @wesmason @kylecrawshaw I believe this is ready to merge. Any other feedback?

kylecrawshaw commented:
@gsong I'll set aside some time today to take a look at it from the perspective of IT.

nedbat commented:
@gsong I had some outstanding review comments, though GitHub isn't making it easy to find them....

gsong commented:
@nedbat I'll take that as a 👍?

nedbat commented:
:+1:
